server/scsynth/SC_PulseAudio.cpp

```diff
@@ -0,0 +1,320 @@
+        mPeakCPU = cpuUsage;
+        mPeakCounter = mMaxPeakCounter;
+    }
+
+#endif
+    mAudioSync.Signal();
+    return 0;
+}
+
+
+// ========================================================================
+// The SC_AudioDriver interface methods that we need to implement are below
+
+bool SC_PulseAudioDriver::DriverSetup(int* outNumSamples, double* outSampleRate) {
+    int rc;
+    int device;
+    scprintf("In SC_PulseAudioDriver::DriverSetup\n");
+
+    // Show the devices
+    int numDevices = m_audio->getDeviceCount();
+    if (numDevices == 0) {
+        scprintf("No audio devices found\n");
+        return false;
+    }
+
+    scprintf("Found %d device(s):\n", numDevices);
+    std::vector<RtAudio::DeviceInfo> infos;
+    for (int i = 0; i < numDevices; i++) {
+        try {
+            RtAudio::DeviceInfo info;
+            info = m_audio->getDeviceInfo(i);
+            scprintf("  - %s   (device #%d with %d ins %d outs)\n", info.name.c_str(), i, info.inputChannels,
+                     info.outputChannels);
+
+            infos.push_back(info);
+        } catch (RtAudioError& e) {
+            e.printMessage();
+            break;
+        }
+    }
+
+    // Use the default devices
+    unsigned int outputDevice = m_audio->getDefaultOutputDevice();
+    unsigned int inputDevice = m_audio->getDefaultInputDevice();
+
+    // Number of input and output channels to use
+    m_outputChannelCount = std::min<size_t>(mWorld->mNumOutputs, infos[outputDevice].duplexChannels);
+    m_inputChannelCount = std::min<size_t>(mWorld->mNumInputs, infos[inputDevice].duplexChannels);
+
+    // What sample rate to use? If we get one, make sure it is supported, and fall back to the default if not
+    *outNumSamples = mWorld->mBufLength;
+    if (mPreferredSampleRate) {
+        std::vector<unsigned int> supportedSampleRates { infos[outputDevice].sampleRates };
+        if (std::find(supportedSampleRates.begin(), supportedSampleRates.end(), (unsigned int)mPreferredSampleRate)
+            != supportedSampleRates.end()) {
+            *outSampleRate = mPreferredSampleRate;
+        } else {
+            scprintf("Requested sample rate NOT supported. Setting to 44.1 kHz\n");
+            *outSampleRate = 44100;
+        }
+    } else {
+        *outSampleRate = 44100;
+    }
+
+    // Configure the stream parameters
+    RtAudio::StreamParameters outParameters;
+    outParameters.deviceId = outputDevice;
+    outParameters.nChannels = m_outputChannelCount;
+    RtAudio::StreamParameters inParameters;
+    inParameters.deviceId = inputDevice;
+    inParameters.nChannels = m_inputChannelCount;
+    // use a separate variable for the number of frames, so that we can compare with the actual given value
+    // and warn if they are different
+    unsigned int bufferFrames = *outNumSamples;
+    RtAudio::StreamOptions options;
+    options.flags = RTAUDIO_MINIMIZE_LATENCY | RTAUDIO_SCHEDULE_REALTIME;
+    options.streamName = "SuperCollider";
+    try {
```

Comment:
In this block, we should try to determine the driver's latency and print stats (sample rate, hardware block size, latency etc.).

Reply:
I need to check and see what information we can get back regarding latency. I think I am already printing some stats somewhere else, I'll have a look.

Reply:
I have added output about the sample rate and the hardware block size. I did not find any way to report the latency.

Reply:
Great! Should be able to test later today.

IIUC audio latency = buffer size / sample rate.

Also, set `m_maxOutputLatency` to this value (with appropriate casting -- actually I'm not sure of it, seems to be in seconds but could be something else). That is likely to improve the perceived latency I got with timestamped OSC messages.

Reply:
Definitely a few issues here.

The hardware buffer size is *not* `bufferFrames`. Compare to the PortAudio backend, which uses `mPreferredHardwareBufferFrameSize`.

`bufferFrames` is the ServerOptions `blockSize`, which determines control rate and is intentionally smaller than the hardware buffer.

The result is, with default settings, RtAudio is trying to initialize to a 64-sample hardware buffer (1.45 ms). This is of course the opposite of the intention here: we wanted to trade latency for ease of configuration, but instead end up with an extremely low latency configuration that may not behave well.

This influences CPU readings as well: average % is quite low, half a percent or less, but the peak is around 3%. When I started the server with a 256-sample control block, suddenly the CPU readings stabilized (but the primary determinant for CPU performance here should be the hardware size, not the control block size -- and this means kr UGens will run unacceptably slow).

To get acceptable performance, the RtAudio/PulseAudio backend needs to handle the ServerOptions `hardwareBufferSize` and `blockSize` settings in the same way that the CoreAudio and PortAudio backends do.

Reply:
Thanks for clarifying the purpose of the different variables, as you can see it is an area that I was struggling to really understand. I'll fix the code with this new guidance.

Regarding calculating latency, yes, I understand that the latency at this end can be calculated that way, but I believe that the PulseAudio server itself can add some latency itself as well.

Reply:
@jamshark70, I think that with the latest commit on Friday, now we are handling properly the difference between block size and the hardware buffer size. It would be great if you can have a look when you have the chance.

Reply:
@jamshark70, did you have the chance to have a look at the changes in this area? I am very interested in hearing back from you about this. I believe that the current code manages the block size and the hardware buffer size properly, but would like confirmation. Thanks!

Reply:
FWIW:
- PortAudio implementation reports latency for input and output independently
  - not sure whether that parameter takes into account additional system processing, although experiential tests of PortAudio on macOS did yield comparable results between reported and actual latency
- CoreAudio implementation does not report latency, only buffer size
- Jack implementation reports `max output latency`, but I don't know what that parameter means actually

@jamshark70 @llloret do you think there's a point in reporting latency if we cannot really approximate total latency?

Finally, what do you think about printing `buffer size` as opposed to `block size`? Indeed CoreAudio prints `block size` but IMO this is easily confused with the "processing block size" parameter (numer of audio samples per control period) and I think using `buffer size` when talking about soundcard buffer would be better. It also matches better with sclang's `hardwareBufferSize`.

Reply:
> Finally, what do you think about printing buffer size as opposed to block size?

I just committed this.

Reply:
> jamshark70, did you have the chance to have a look at the changes in this area? I am very interested in hearing back from you about this. I believe that the current code manages the block size and the hardware buffer size properly, but would like confirmation.

I haven't reviewed the code in depth, but the behavior seems to be right. I also tested setting hardware buffer size and sample rate explicitly, and it seemed to take effect as it should. (It should have a more detailed review from a C++ guru though.)